### ISSUE TITLE
Preprocessing laziness and the Corpus inheritance hierarchy

### ISSUE THREAD

**Issue by oadams** [oadams/persephone#98]
The current way preprocessing of corpora is addressed is by making `corpus.Corpus` and its subclasses address preprocessing in their `__init__()` methods. For use this approach to be feasible, this initialization needs to be lazy:

1. `corpus.Corpus.prepare_feats()` only prepares features if they don't exist.
2.  `corpus.Corpus.make_data_splits()` is only run if no `valid_prefixes.txt` or `test_prefixes.txt` files are present.
3. `preprocess.elan.Corpus.__init__()` only digs `Utterance`s out of ELAN files and extracts wavs from the corresponding media files if the `tgt_dir` doesn't exist.

There's a couple different granularities for the laziness in the current approach. Examples 1-2 above have quite a fine granularity: they check for each file at the utterance level, so if preprocessing was previously cancelled halfway through, they should pick up where they left off. Example 3 currently just checks if a target directory exists, which is a bit dubious and means the code won't preprocess data even if previous preprocessing didn't happen for some reason or was incomplete. One way to solve this is to offer an argument to the constructor, `preprocess.elan.Corpus(reset=True)`, which deletes the target data directory and performs preprocessing again.

If these corpora weren't lazy, then creating a `Corpus` object would take forever all the time. This would obviously be very bad. On the other hand, creating `Corpus` objects with the current approach can take a wildly variable amount of time. An alternative might be to specifically request preprocessing via the interface, with something like:
```
some_corpus = corpus.Corpus("/some/dir")
some_corpus.preprocess()
```
How best then to handle the unpreprocessed Corpus object? Does `corpus_reader.CorpusReader` then have to check if a `Corpus` has an `preprocessed = False` flag?

A closely related but distinct issue is that  any upstream changes to the original data currently cannot be accounted for. One simple idea is to take a checksum on all the original data, and store that in the `tgt_dir` (where the preprocessed data is kept). When a new Corpus object is created and it determines that preprocessing has already happened, it can compare this checksum against a checksum in the original data directory. If it has changed then it can fire off a warning.

**Comment by shuttle1987** [comment]
A few weeks ago we talked about some design ideas on the phone. This conversation I remember mentioning that I thought that this particular area of the code would be benefited from a pipeline of processes that get run. It's the hidden state that bothers me here, you have these different operations happening behind the scenes that could have substantial differences in runtime based on some state that may not be visible.

One possible approach is to take a hash of all the files and then check for file changes if you want to reuse the results. This way you only reuse results if that hash matches the old hash of the files and can then save some error states where the files have changed since the cached results were computed.

**Comment by oadams** [comment]
@alexis-michaud Thanks for sharing your perspective on this as the linguist and user. As always, it's good to get that viewpoint. My conclusions from your comments:
- Addressing the upstream changes to the data is a key issue. The solution I mentioned with checksums would resolve this, and such that there should be no need to re-run unnecessary preprocessing, not run the the risk of upstream changes not being noticed.
- In a similar spirit, your comments suggest the solution to the laziness granularity problem is to look at each preprocessed file individually, and not just check for the existence of a target directory. (I was leaning this way anyway).

> Said differently, for novice users I'd suggest setting the Warning  the other way round. If the data are read from the linguist's files  (=doing all the preprocessing): no warning (this is standard). If the data are inherited from the preprocessed files (I'll call them  "auxiliary files"), then a warning should be issued: "This data set was  already preprocessed. Data are inherited from previous preprocessing. To  purge the cache and start fresh from the input files, ... (provide  instructions)"

So to be clear, the preprocessed data would be what the user has already supplied previously. If we can guarantee that no change has happened to the input data, then there shouldn't be a need to present a warning unless the data has been changed. Given your thoughts, the best thing here is to just automatically re-run the preprocessing if the original data changes.

@shuttle1987 So I actually had that discussion we had in mind while thinking about this! My thoughts were that one way to go for describing a pipeline is to describe the `Corpus` preprocessing class (say `preprocess.Elan.Corpus`), it's arguments, and the arguments of the `CorpusReader` and `Model` to train. When the "Run" button is clicked, those objects are actually instantiated and model.train() is called.

The main alternative to that approach I can see would be for `Corpus.__init__()` to not do preprocessing by default, and require the user to call a `Corpus.preprocess()` method. The issue here is that it adds complexity to the interface to the objects. The user has to know to call `Corpus.preprocess()`. It means `Corpus` objects can be in two states, requiring further checks when an object is used. Validating whether the `Corpus` would be good to go given the arguments is hard because that requires actually preprocessing the data. The advantage of getting the `Corpus` object created immediately doesn't seem to make up for the added interface complication. With the current approach, once `Corpus.__init__()` is the user of the can be sure the Corpus is ready. Any existent `Corpus` object is an actual corpus that can be used.

As for the hashing, I'm on board and I'm more and more convinced this is essential. It's a minor increase in complexity but worth it since preprocessing can sometimes be very long. For instance, a dataset I'm working with at the moment can take about an hour to preprocess (extracting WAVs from certain parts of MP4 video being the main bottleneck). But as mentioned above wrt. Alexis's comment, it's more than about speed-up. What's really important is that if a user changes what's in the input directory, the code needs to cope with that when new `Corpus` objects are created, even if preprocessing has already been done previously.

To me there is one minor question about the hashing. As is planned, when a `Corpus` is created, the hash of the original data can be stored with the preprocessed data so that when new `Corpus` objects are created they know whether they need to do the preprocessing again. This is clear. But is there a case for having existing `Corpus` objects also check up on the input data too and compare it against the hash they previously stored when certain methods are called? Say if `Corpus.get_train_fns()` is called, it might detect a different hash and warn that the input data has changed. I think this probably isn't worth it. It adds more complexity and overhead than is needed. The hashing check should probs just happen once during object construction. Once a `Corpus` is created, it shouldn't worry about changing itself if input data changes. A goal is to have these `Corpus` objects as immutable after construction as possible. Of course, pumping out a warning isn't inimical to this, but the overhead probably still isn't really worth it.

**Comment by oadams** [comment]
So this issue is becoming quite a monolith, but here are some design decisions regarding the class inheritance hierarchy and object initialization based on a discussion with @shuttle1987 today.

We are in agreement that the corpus preprocessing and validation all happens in `corpus.Corpus.__init__()` and that once that is done, the object is essentially treated as immutable. Experiment pipelining (via a GUI) describes how `Corpus` objects are to be created without creating them.

The current way `Corpus` objects are created based on different preprocessing requirements involves inheritance. Eg. `elan.Corpus` does its own preprocessing in its own `__init__()` and then calls `super().__init__()`. Similarly, `bkw.Corpus` does Kunwinjku-specific preprocessing in `bkw.__init__()` before calling `super().__init__()` (`elan.Corpus.__init__()`). @shuttle1987 made the argument that since only the constructor is changing, a factory is more appropriate here than an inheritance hierarchy. So `corpus.Corpus` would have class method constructors to build `Corpus` objects from ELAN eafs, Praat TextGrids, Pangloss XML files, etc. All these general reusable components would have their own constructors. Dataset-specific stuff, which is currently handled by things like `bkw.Corpus`, wouldn't go in `corpus.Corpus` (of course, the class would become massive before long), but rather `datasets.bkw` would provide its own functionality for creating a `Corpus` object, still without inheritance.

Aside from making the hierarchy flatter and easier to comprehend, this has another nice effect in that there aren't subclasses of `Corpus` spread across the `preprocess` and `datasets` subpackages.

**Comment by oadams** [comment]
The design decisions made from this discussion haven't changed.
